src/thunderbolt/store.c

```diff
@@ -0,0 +1,465 @@
+
+        r = lstat(path, &st);
+        if (r < 0 && errno == ENOENT) {
+                ret->level = AUTH_MISSING;
+                return 0;
+        }
+        if (r < 0)
+                return -errno;
+        if (S_ISREG(st.st_mode)) {
+                _cleanup_free_ char *l = NULL;
+
+                r = read_one_line_file(path, &l);
+                if (r < 0)
+                        return r;
+
+                ret->store = STORE_FSDB;
+                if (strlen(l) == KEY_CHARS) {
+                        ret->level = AUTH_KEY;
+                        ret->key = l;
+                        l = NULL;
+                } else {
+                        ret->key = NULL;
+                        r = safe_atoi(l, &ret->level);
+                }
+                return r;
+        } else if (!S_ISLNK(st.st_mode)) {
+                return -ENOTSUP;
+        }
+
+        r = readlink_malloc(path, &p);
+        if (r < 0)
+                return r;
+
+        r = -ENOTSUP;
+        if (startswith(p, "/sys/firmware/efi/efivars")) {
+                r = store_efivars_get_auth(uuid, ret);
+        }
+
+        if (r == -ENOTSUP) {
+                ret->level = AUTH_MISSING;
+                ret->store = STORE_NONE;
+        }
+
+        return r;
+}
+
+#define ID128_UUID_FMT "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x"
+
+#define TB_EFIVAR_PATH_PREFIX "/sys/firmware/efi/efivars/Thunderbolt-"
+#define TB_EFIVAR_PATH TB_EFIVAR_PATH_PREFIX ID128_UUID_FMT
+
+static int store_efivars_put_auth(TbStore *store,
+                                  const char *uuid,
+                                  Auth *auth) {
+        _cleanup_free_ char *target = NULL;
+        char buf[FORMAT_SECURITY_MAX];
+        sd_id128_t id;
+        char *path;
+        int r;
+
+
+        if (sd_id128_from_string (uuid, &id) < 0) {
+                return -EINVAL;
+        }
+
+        if (auth->level == AUTH_KEY) {
+                r = efi_set_variable(id, "Thunderbolt", auth->key, KEY_CHARS);
+        } else {
+                xsprintf(buf, "%hhu", (uint8_t) auth->level);
+                r = efi_set_variable(id, "Thunderbolt", buf, 1);
+        }
+
+        if (r < 0)
+                return r;
+
+        if (asprintf(&target, TB_EFIVAR_PATH, SD_ID128_FORMAT_VAL(id)) < 0)
+                 return -ENOMEM;
+
+        path = strjoina(store->path, "/authorization/", uuid);
+
+        r = mkdir_parents(path, 0755);
+        if (r < 0)
+                return r;
+
+        return symlink_idempotent(target, path);
+}
+
+
+static int store_fsdb_put_auth(TbStore *store,
+                               const char *uuid,
+                               Auth *auth) {
+        char buf[KEY_CHARS + 1];
+        char *path;
+        int r;
+
+        if (auth->level == AUTH_KEY) {
+                xsprintf(buf, "%s", auth->key);
+        } else {
+                xsprintf(buf, "%hhu", (uint8_t) auth->level);
```

Comment:
Hm, doesn't that get converted right back to an `int` anyway? I.e. you might drop the cast without any effect.